cosigner/authcosigner.go

```diff
@@ -0,0 +1,224 @@
+package cosigner
+
+import (
+	"crypto"
+	"crypto/hmac"
+	"crypto/rand"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	"github.com/lestrrat-go/jwx/v2/jws"
+	"github.com/openpubkey/openpubkey/cosigner/msgs"
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type AuthCosigner struct {
+	Cosigner
+	Issuer       string
+	KeyID        string
+	authIdIter   atomic.Uint64
+	hmacKey      []byte
+	AuthStateMap map[string]*AuthState
+	AuthCodeMap  map[string]string
+}
+
+func New(signer crypto.Signer, alg jwa.SignatureAlgorithm, issuer, keyID string) (*AuthCosigner, error) {
+	hmacKey := make([]byte, 64)
+	if _, err := rand.Read(hmacKey); err != nil {
+		return nil, err
+	}
+
+	return &AuthCosigner{
+		Cosigner: Cosigner{
+			alg:    alg,
+			signer: signer},
+		Issuer:       issuer,
+		KeyID:        keyID,
+		authIdIter:   atomic.Uint64{},
+		hmacKey:      hmacKey,
+		AuthStateMap: make(map[string]*AuthState),
+		AuthCodeMap:  make(map[string]string),
+	}, nil
+}
+
+func (c *AuthCosigner) InitAuth(pkt *pktoken.PKToken, sig []byte) (string, error) {
+	msg, err := pkt.VerifySignedMessage(sig)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse sig: %w", err)
+	}
+	var initMFAAuth *msgs.InitMFAAuth
+	if err := json.Unmarshal(msg, &initMFAAuth); err != nil {
+		return "", fmt.Errorf("failed to parse InitMFAAuth message: %w", err)
+	} else if time.Since(time.Unix(initMFAAuth.TimeSigned, 0)).Minutes() > 2 {
+		return "", fmt.Errorf("timestamp (%d) in InitMFAAuth message too old, current time is (%d)", initMFAAuth.TimeSigned, time.Now().Unix())
+	} else if time.Until(time.Unix(initMFAAuth.TimeSigned, 0)).Minutes() > 2 {
+		return "", fmt.Errorf("timestamp (%d) in InitMFAAuth message too far in the future, current time is (%d)", initMFAAuth.TimeSigned, time.Now().Unix())
+	} else if authState, err := NewAuthState(pkt, initMFAAuth.RedirectUri, initMFAAuth.Nonce); err != nil {
+		return "", err
+	} else if authID, err := c.CreateAuthID(uint64(time.Now().Unix())); err != nil {
+		return "", err
+	} else {
+		c.AuthStateMap[authID] = authState
+		return authID, nil
+	}
+}
+
+func (c *AuthCosigner) CreateAuthID(timeNow uint64) (string, error) {
+	authIdInt := c.authIdIter.Add(1)
+	iterAndTime := []byte{}
+	iterAndTime = binary.LittleEndian.AppendUint64(iterAndTime, uint64(authIdInt))
+	iterAndTime = binary.LittleEndian.AppendUint64(iterAndTime, timeNow)
+	mac := hmac.New(crypto.SHA3_256.New, c.hmacKey)
+	if n, err := mac.Write(iterAndTime); err != nil {
+		return "", err
+	} else if n != 16 {
+		return "", fmt.Errorf("unexpected number of bytes read by HMAC, expected 16, got %d", n)
+	} else {
+		return hex.EncodeToString(mac.Sum(nil)), nil
+	}
+}
```

Comment:
I avoided using a UUID here because they only have 122-bits, so under the birthday bound, they only have 64-bits of collision resistance. As [wikipedia says](https://en.wikipedia.org/wiki/Universally_unique_identifier) to get a 50% chance of collision with a UUIDv4 you would need to generate "1 billion UUIDs per second for about 86 years", so 122-bits is probably safe.

Then again [the OAuth2 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-10.10) hates on the idea of anything less than 128-bit collision resistance:

>  The probability of an attacker guessing generated tokens (and other credentials not intended for handling by end-users) MUST be less than or equal to 2^(-128) and SHOULD be less than or equal to 2^(-160).

I just didn't want to have to have to justify why 64-bit is "good enough" in this case. So I choose a 256-bit (16-byte) ID giving me 128-bits of collision resistance. Most of OpenPubkey is >=256-bit values so anyone doing a security analysis so it aligns well with the other cryptographic values.